Mathlib/RingTheory/TensorProduct/MonoidAlgebra.lean

```diff
@@ -0,0 +1,228 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.LinearAlgebra.DirectSum.Finsupp
+import Mathlib.Algebra.MonoidAlgebra.Basic
+import Mathlib.RingTheory.TensorProduct.Basic
+import Mathlib.Algebra.Algebra.Equiv
+import Mathlib.Data.Finsupp.Defs
+
+/-! # TensorProduct of a monoid algebra
+
+This file consists of two sections, the first one will have to be moved
+somewhere else.
+
+## functoriality for `MonoidAlgebra`
+
+We prove some functoriality definitions for the `Semiring` variable.
+It remains to prove that this is a functor.
+
+## tensor product
+
+* `MonoidAlgebra.rTensorAlgEquiv`, the tensor product of
+   `MonoidAlgebra M α` with `N` is `R`-linearly equivalent to `MonoidAlgebra (M ⊗[R] N) α`
+
+* `MonoidAlgebra.scalarRTensorAlgEquiv`, the tensor product of
+   `MonoidAlgebra R α` with `N` is `R`-linearly equivalent to `MonoidAlgebra N α`
+
+## TODO
+
+* State and prove functoriality properties
+
+* Prove the `lTensor` variants
+
+* Combine both to get that the tensor product of `MonoidAlgebra M α`
+  with `MonoidAlgebra N β` is `R`-linearly equivalent to
+  `MonoidAlgebra (M ⊗[R] N) (α × β)`.
+-/
+
+variable {α R M N P : Type*}
+
+section Functoriality
+
+namespace MonoidAlgebra
+
+variable [DecidableEq α]
+
+open Finsupp
+
+variable [Semiring N] [Semiring P]
+
+lemma apply_single
+    (e : N →+ P) (a : α) (n : N) (b : α) :
+    e ((single a n) b) = single a (e n) b :=
+  Finsupp.apply_single e a n b
+
+/-- RingHom functoriality for the monoid algebra -/
+noncomputable def ringHom [MulOneClass α] (e : N →+* P) :
+    MonoidAlgebra N α →+* MonoidAlgebra P α := {
```

Comment:
I think this might already exists; something like `MonoidAlgebra.liftNC` or similar?

Reply:
docs#MonoidAlgebra.liftNCRingHom does take into account the multiplicative structure, but uses `Monoid`, while here `MulOneClass` is enough. What should one do? (I am not sure that docs#MonoidAlgebra.liftNCRingHom can be generalized, docs#MonoidAlgebra.liftNC_mul does not suffice, apparently.)

Reply:
At least, one can give the shorter (but less general) definition 
```
/-- RingHom functoriality for the monoid algebra -/
noncomputable def ringHom [Monoid α] (e : N →+* P) :
    MonoidAlgebra N α →+* MonoidAlgebra P α := by
  apply liftNCRingHom (singleOneRingHom.comp e) (of P α)
  intro n a
  simp only [commute_iff_eq, RingHom.coe_comp, Function.comp_apply,
    singleOneRingHom_apply, ZeroHom.toFun_eq_coe, AddMonoidHom.toZeroHom_coe,
    singleAddHom_apply, of_apply, single_mul_single, one_mul, mul_one]
```
and I can try proving the lemmas for them. (This is not as obvious, I need to understand why!)

Reply:
That works, but one loses the fact that it corresponds with docs#Finsupp.mapRange…
In the current version, there is a proof of everything but for one definition, 
and proofs/definitions exist in two ways, to help you decide which way is better.